README.md
# myhabeetat
Wrapper for MyHabeetat API, control your devices using nodejs.

Supports Solidmation devices, including: 
- BGH Smart Control

## Installation
Install it via NPM:

```bash
npm install --save myhabeetat
```


## Usage

```javascript

const MyHabeetat = require('myhabeetat')

// Get auth token
let token = await MyHabeetat.login('myawesome@email.com', 's3cret_passw0rd')

// Get homes and devices
let homes = await MyHabeetat.getHomes(token)
let devices = await MyHabeetat.getDevices(token, homes[0])

// Turn on device. 15 is model code for BGH Smart Control
let result = await MyHabeetat.turnOnDevice(token, 15, devices[0].endpoint)
```

## MyHabeetat methods

### MyHabeetat.login(email, password)
#### Arguments
- email (string): Your My Habeetat login email
- password (string): Your My Habeetat login password 

#### Returns
- token (string): Authentication token for My Habeetat's API

### MyHabeetat.getHomes(token)
#### Arguments
- token (string): A valid My Habeetat auth token

#### Returns
- homes (array): An array of home IDs

### MyHabeetat.getDevices(token, home)
#### Arguments
- token (string): A valid My Habeetat auth token
- home (integer): Home ID to get devices from

#### Returns
- devices (array): Array of device objects for the given home

### MyHabeetat.getDeviceStatus(token, home, device)
#### Arguments
- token (string): A valid My Habeetat auth token
- home (integer): Home ID to get devices from
- device (integer): Device ID to get status from

#### Returns
- status (object): A status object for the given device

### MyHabeetat.setDeviceStatus(token, model, endpoint, status)
#### Arguments
- token (string): A valid My Habeetat auth token
- model (integer): Device model ID. Currently only BGH Smart Control is supported (ID=15)
- endpoint (integer): Endpoint ID for the device. Can be retrieved with `getDeviceStatus` method

#### Returns
- result (boolean): True if the operation succeded.

### MyHabeetat.getDataPacket(token, home)
#### Arguments
- token (string): A valid My Habeetat auth token
- home (integer): Home ID to get devices from

#### Returns
- data (object): Last data packet for the device


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v1.5.0
## (2020-07-13)

* device: add support for BGH Silent Air [Tomás Migone]

# v1.4.3
## (2020-07-13)

* meta: sync package versions [Tomás Migone]

# v1.4.2
## (2020-07-12)

* meta: add versionist action [Tomás Migone]
* Added README.md [Tomás Migone]


tests/bgh.test.js
/* eslint-disable no-unused-expressions */
const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')
chai.use(chaiAsPromised)
const expect = chai.expect

const BGHSmartControl = require('../lib/BGHSmartControl.js')

let goodEndpoint = [
  { Value: '24.3', ValueType: 13 },
  { Value: '2', ValueType: 14 },
  { Value: '3', ValueType: 15 },
  { Value: '0', ValueType: 16 },
  { Value: '0', ValueType: 17 },
  { Value: '0', ValueType: 18 },
  { Value: '26', ValueType: 20 },
  { Value: '77', ValueType: 59 },
  { Value: '4170', ValueType: 60 },
  { Value: '86', ValueType: 61 },
  { Value: '0', ValueType: 62 }
]
let badEndpoint = []
let goodEndpointResults = {
  mode: 'HEAT',
  fanMode: 'HIGH',
  temperature: 24.3,
  targetTemperature: 26
}

describe('#BGHSmartControl.parse()', function () {
  context('with good endpoint values', function () {
    let data
    before(function () {
      data = BGHSmartControl.parseStatus(goodEndpoint)
    })
    it('should return the parsed data object', function () {
      expect(data).to.be.an('object')
    })
    it('parsed object should have all BGH status variables', function () {
      expect(data).to.have.keys(['mode', 'fanMode', 'temperature', 'targetTemperature'])
    })
    it('BGH status variables should match their expected value', function () {
      expect(data.mode).to.equal(goodEndpointResults.mode)
      expect(data.fanMode).to.equal(goodEndpointResults.fanMode)
      expect(data.temperature).to.equal(goodEndpointResults.temperature)
      expect(data.targetTemperature).to.equal(goodEndpointResults.targetTemperature)
    })
  })
  context('with bad endpoint values', function () {
    it('should fail and throw a known error', function () {
      expect(() => { BGHSmartControl.parseStatus(badEndpoint) }).to.throw(Error, 'Error parsing packet for device.')
    })
  })
})


tests/homes.test.js
/* eslint-disable no-unused-expressions */
require('dotenv').config()
const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')
chai.use(chaiAsPromised)

const expect = chai.expect
const MyHabeetat = require('../lib/MyHabeetat.js')

let validCredentials = { email: process.env.USER_VALID, password: process.env.PASSWORD_VALID }
let invalidToken = 'definitely-not-a-valid-token'

describe('#MyHabeetat.getHomes()', function () {
  this.timeout(5000) // Solidmation's API is really slow
  context('with valid token and existing homes', function () {
    let token = ''
    before(async () => {
      token = await MyHabeetat.login(validCredentials.email, validCredentials.password)
    })
    it('should return an array with the home ids', async () => {
      let homes = await MyHabeetat.getHomes(token)
      expect(homes).to.be.an('array')
    })
  })
  context('with invalid token', function () {
    it('should fail and throw a known error', async () => {
      await expect(MyHabeetat.getHomes(invalidToken)).to.be.rejectedWith(Error, 'Provided token is not valid.')
    })
  })
})


tests/login.test.js
/* eslint-disable no-unused-expressions */
require('dotenv').config()
const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')
chai.use(chaiAsPromised)

const expect = chai.expect
const MyHabeetat = require('../lib/MyHabeetat.js')

let invalidCredentials = { email: 'fake@email.com', password: '123456' }
let validCredentials = { email: process.env.USER_VALID, password: process.env.PASSWORD_VALID }

describe('#MyHabeetat.login()', function () {
  context('with valid credentials', function () {
    it('should return a valid token', async () => {
      let token = await MyHabeetat.login(validCredentials.email, validCredentials.password)
      expect(token).to.be.a('string').that.is.not.empty
    })
  })
  context('with invalid credentials', function () {
    it('should reject login attempt', async () => {
      await expect(MyHabeetat.login(invalidCredentials.email, invalidCredentials.password)).to.be.rejectedWith(Error, 'Provided credentials are not valid')
    })
  })
})


lib/MyHabeetat.js
const config = require('../config/default.js')
const axios = require('axios')
const BGHSmartControl = require('./BGHSmartControl.js')

// Supported device models
const DEVICE_MODELS = {
  BGH_SMART_CONTROL: 15,
  BGH_SILENT_AIR: 10
}

async function login (email, password) {
  let { data } = await axios.post(config.urls.login, { AccountEmail: email, Password: password })

  // Wrong user/pass
  if (data.AccessToken === '') {
    throw new Error('Provided credentials are not valid.')
  }

  return data.AccessToken
}

async function getHomes (token) {
  let { data } = await axios.post(config.urls.enumHomes, { token: { Token: token } })
  let { Homes, ResponseStatus } = data.EnumHomesResult

  // Status = 2 appears to be for generic errors
  if (ResponseStatus.Status === 2) {
    throw new Error('Provided token is not valid.')
  }

  // Homes is null if there are none registered
  return Homes ? Homes.map(h => h.HomeID) : []
}

async function getDataPacket (token, home) {
  let { data } = await axios.post(config.urls.getDataPacket, {
    homeID: home,
    token: { Token: token },
    timeOut: 10000,
    serials: { Home: 0, Groups: 0, Devices: 0, Endpoints: 0, EndpointValues: 0, Scenes: 0, Macros: 0, Alarms: 0 }
  })

  // Status = 2 appears to be for generic errors
  if (data.GetDataPacketResult.ResponseStatus.Status === 2) {
    throw new Error('Provided token or home ID not valid.')
  }

  return data.GetDataPacketResult
}

async function getDevices (token, home) {
  let packet = await getDataPacket(token, home)

  // Devices is null if there are none registered
  return packet.Devices ? packet.Devices.map(d => ({
    id: d.DeviceID,
    home: d.HomeID,
    model: d.DeviceModelID,
    name: d.Description,
    endpoint: packet.Endpoints.find(e => e.DeviceID === d.DeviceID).EndpointID
  })) : []
}

async function getDeviceStatus (token, home, device) {
  let _device, endpoint, data
  let packet = await getDataPacket(token, home)

  // Get device info
  try {
    _device = packet.Devices.find(d => d.DeviceID === device)
  } catch (error) {
    throw new Error('Provided device ID is invalid.')
  }

  // Get device data
  try {
    let endpointID = packet.Endpoints.find(e => e.DeviceID === device).EndpointID
    endpoint = packet.EndpointValues.find(e => e.EndpointID === endpointID)
  } catch (error) {
    throw new Error('Error getting device data.')
  }

  switch (_device.DeviceModelID) {
    case DEVICE_MODELS.BGH_SMART_CONTROL:
    case DEVICE_MODELS.BGH_SILENT_AIR:
      data = BGHSmartControl.parseStatus(endpoint.Values)
      break
    default:
      throw new Error('Device not supported.')
  }

  return data
}

async function setDeviceStatus (token, model, endpoint, status) {
  let success
  switch (model) {
    case DEVICE_MODELS.BGH_SMART_CONTROL:
    case DEVICE_MODELS.BGH_SILENT_AIR:
      success = await BGHSmartControl.setStatus(token, endpoint, status)
      break
    default:
      throw new Error('Device not supported.')
  }

  return success
}

async function turnOnDevice (token, model, endpoint) {
  let success
  switch (model) {
    case DEVICE_MODELS.BGH_SMART_CONTROL:
    case DEVICE_MODELS.BGH_SILENT_AIR:
      success = await BGHSmartControl.turnOn(token, endpoint)
      break
    default:
      throw new Error('Device not supported.')
  }

  return success
}

async function turnOffDevice (token, model, endpoint) {
  let success
  switch (model) {
    case DEVICE_MODELS.BGH_SMART_CONTROL:
    case DEVICE_MODELS.BGH_SILENT_AIR:
      success = await BGHSmartControl.turnOff(token, endpoint)
      break
    default:
      throw new Error('Device not supported.')
  }

  return success
}

module.exports = {
  login,
  getHomes,
  getDevices,
  getDataPacket,
  getDeviceStatus,
  setDeviceStatus,
  turnOnDevice,
  turnOffDevice
}


lib/BGHSmartControl.js
const config = require('../config/default.js')
const axios = require('axios')

const MODES = [
  { name: 'OFF', code: 0 },
  { name: 'COOL', code: 1 },
  { name: 'HEAT', code: 2 },
  { name: 'DRY', code: 3 },
  { name: 'FAN', code: 4 },
  { name: 'AUTO', code: 254 },
  { name: 'NO_CHANGE', code: 255 }
]
const FAN_MODES = [
  { name: 'SLOW', code: 1 },
  { name: 'MID', code: 2 },
  { name: 'HIGH', code: 3 },
  { name: 'AUTO', code: 254 },
  { name: 'NO_CHANGE', code: 255 }

]
const CODES = { TEMPERATURE: 13, TARGET_TEMPERATURE: 20, FAN_MODE: 15, MODE: 14 }

async function setStatus (token, endpoint, status) {
  let mode = MODES.find(m => m.name === status.mode)
  let fanMode = FAN_MODES.find(m => m.name === status.fanMode)
  let { data } = await axios.post(config.urls.HVACSetModes, {
    token: { Token: token },
    flags: 255,
    endpointID: endpoint,
    mode: mode ? mode.code : MODES.find(m => m.name === 'NO_CHANGE').code,
    fanMode: fanMode ? fanMode.code : FAN_MODES.find(m => m.name === 'NO_CHANGE').code,
    desiredTempC: status.targetTemperature || -327.68 // Random? Web panel uses that number for some reason for target
  })
  // Return true if command applied successfully, false if not
  return (data.HVACSetModesResult && data.HVACSetModesResult.ResponseStatus && data.HVACSetModesResult.ResponseStatus.Status === 0) || false
}

async function turnOn (token, endpoint) {
  let success = await setStatus(token, endpoint, {
    mode: 'AUTO'
  })
  return success
}

async function turnOff (token, endpoint) {
  let success = await setStatus(token, endpoint, {
    mode: 'OFF'
  })
  return success
}

function parseStatus (data) {
  let temperature, targetTemperature, fanMode, mode
  try {
    // Current temperature
    temperature = parseFloat(data.find(v => v.ValueType === CODES.TEMPERATURE).Value)

    // Target temperature
    targetTemperature = parseFloat(data.find(v => v.ValueType === CODES.TARGET_TEMPERATURE).Value)

    // Fan mode
    let fanCode = parseInt(data.find(v => v.ValueType === CODES.FAN_MODE).Value)
    fanMode = FAN_MODES.find(m => m.code === fanCode).name

    // Unit mode
    let modeCode = parseInt(data.find(v => v.ValueType === CODES.MODE).Value)
    mode = MODES.find(m => m.code === modeCode).name
  } catch (error) {
    throw new Error(`Error parsing packet for device: ${error.message}`)
  }

  return {
    mode,
    fanMode,
    temperature,
    targetTemperature
  }
}

module.exports = {
  parseStatus,
  setStatus,
  turnOn,
  turnOff
}
